R/weighted_efficacy.R

```diff
@@ -0,0 +1,275 @@
+#' @param prioritisation_matrix Prioritisation matrix
+#' @param d2_prioritise Boolean vector indicating which age groups are prioritised to receive 2nd dose
+#' @param infection_efficacy Vector of length 2 of infection efficacy for first and second dose
+#' @param disease_efficacy Vector of length 2 of disease efficacy for first and second dose
+#'
+#' @return Age disaggregated dose numbers and efficacies over time.
+#' @export
+weighted_efficacy <- function(iso3c,
+                              N,
+                              maxt,
+                              doses_per_day,
+                              dose_period,
+                              v1v2,
+                              prioritisation_matrix,
+                              d2_prioritise,
+                              infection_efficacy,
+                              disease_efficacy){
+
+  stopifnot(v1v2 < dose_period)
+  # Get population
+  pop <- squire::get_population(iso3c = iso3c)
+  # Rescale population and define age group for each individual
+  ages <- rep(1:nrow(pop), round(N * (pop$n / sum(pop$n))))
+  # Record of when dose 1 and 2 were given for each individual
+  dose_times <- data.frame(dose1_t = rep(NA, length(ages)), dose2_t = rep(NA, length(ages)))
+  dose_times <- split(dose_times, ages)
+  # Start with phase 1
+  phase <- 1
+  # Start at first row of prioritisation matrix
+  p_step <- 1
+  for(t in 1:maxt){
+    # Available doses to distribute
+    current_doses <- doses_per_day[t]
+
+    # Phase 1
+    ## For each row of the prioritisation matrix satisfy:
+    ### All first dose coverage >= prioritisation matrix target
+    ### All second dose coverage for age groups prioritised for second doses >= prioritisation matrix target
+    if(phase == 1){
+      # Individuals left to target in each age group
+      tp <- target_pop(dose_number = 1, dose_times = dose_times, prioritisation = prioritisation_matrix[p_step, ], t = t, dose_period = dose_period, d2_prioritise = d2_prioritise)
+      # Assign available doses
+      ad <- assign_doses(current_doses, tp)
+      # Administer available doses
+      dose_times <- administer_first_dose(ad, dose_times, t)
+      # Recalculate remaining doses
+      current_doses <- current_doses - sum(ad)
+
+      if(current_doses > 0){
+        # Individuals left to target in each age group
+        tp <- target_pop(dose_number = 2, dose_times = dose_times, prioritisation = prioritisation_matrix[p_step, ], t = t, dose_period = dose_period, d2_prioritise = d2_prioritise)
+        # Assign available doses
+        ad <- assign_doses(current_doses, tp)
+        # Administer available doses
+        dose_times <- administer_second_dose(ad, dose_times, dose_period, t)
+        # Recalculate remaining doses
+        current_doses <- current_doses - sum(ad)
+      }
+
+      # Any remaining doses can go towards 1st doses in the next prioritisation step
+      if(current_doses > 0){
```

Comment:
Could we change this to a while loop so it explores other steps if need be - from testing you need to go to quite high values of N sometimes to get this to not get stuck. 

```suggestion
      temp_p_step <- p_step
      while(current_doses > 0 & p_step != nrow(prioritisation_matrix)){
      temp_p_step <- temp_p_step + 1
```

Reply:
And by stuck I mean, I was comparing in the below what the empirical data would be say for Bangladesh using the actual values (this is the first section that yields gg1) vs what this function would give me for say N of 100000 and seeing it has periods of not giving out any doses. 

```

# example data
owid <- read.csv("https://covid.ourworldindata.org/data/owid-covid-data.csv")
iso3c <- "BGD"

# data first vaccination given out
date_first_vacc <- owid %>% filter(iso_code == iso3c) %>%
  select(date, people_vaccinated) %>%
  na.omit %>%
  pull(date) %>%
  as.Date() %>%
  min()

# first doses interpolating over missing NAs
first_doses <- owid %>% filter(iso_code == iso3c) %>%
  pull(people_vaccinated) %>%
  zoo::na.approx()

first_doses <- c(first_doses[1], diff(first_doses))

# second doses interpolating over missing NAs
total_doses <- owid %>% filter(iso_code == iso3c) %>%
  pull(total_vaccinations) %>%
  zoo::na.approx()
total_doses <- c(total_doses[1], diff(total_doses))

second_doses <- total_doses - first_doses

dose_table <- function(prioritisation_matrix, pop){

# what are the other inputs
current_cov <- rep(0, 17)
current_2_cov <- rep(0, 17)

# doses per day object
dpd <- vector("list", length = length(first_doses))
p_step_1 <- 1
p_step_2 <- 1

# blank doses data.frame
bdpd <- data.frame("t" = 1, "age_group" = as.factor(1:17), "n_1dose" = 0, "n_2dose" = 0)
for(i in seq_along(dpd)) {

  # first doses first

  # to be given out today if possible
  to_give_out <- (pop$n*prioritisation_matrix[p_step_1,]) - (current_cov)
  total_to_give_out <- sum(to_give_out)

  # first doses first
  if(total_to_give_out > first_doses[i]) {

    going_out <- as.integer((first_doses[i] / total_to_give_out) * to_give_out)

  } else {

    going_out <- to_give_out
    p_step_1 <- p_step_1 + 1

  }

  bdpd$n_1dose <- going_out
  current_cov <- current_cov + going_out

  # second doses first

  # to be given out today if possible
  to_give_out <- (pop$n*prioritisation_matrix[p_step_2,]) - (current_2_cov)
  total_to_give_out <- sum(to_give_out)

  # first doses first
  if(total_to_give_out > second_doses[i]) {

    going_out <- round((second_doses[i] / total_to_give_out) * to_give_out)

  } else {

    going_out <- to_give_out
    p_step_2 <- p_step_2 + 1

  }

  bdpd$n_2dose <- going_out
  current_2_cov <- current_2_cov + going_out

  # and adjust the time
  bdpd$t <- i
  dpd[[i]] <- bdpd


}

dpd_all <- do.call(rbind, dpd)

return(dpd_all)
}

dpd_all <- dose_table(nimue::strategy_matrix("Elderly", max_coverage = 0.8),
                      squire::get_population(iso3c = iso3c))

disease_efficacy <-  c(0.5, 0.5)
infection_efficacy <- c(0.01, 1)
t1 <- nimue::add_weighted_efficacy(dpd_all, infection_efficacy, disease_efficacy, v1v2 = 28)
t1$date <- date_first_vacc + t1$t - 1

g1 <- ggplot(t1 %>% group_by(age_group) %>%
               mutate(n1cum = cumsum(n_1dose),
                      n2cum = cumsum(n_2dose)),
             aes(t, n1cum, col = age_group)) +
  geom_step() + geom_vline(xintercept = c(12*7, 12*7+24), linetype = "dashed")

g2 <- ggplot(t1 %>% group_by(age_group) %>%
               mutate(n1cum = cumsum(n_1dose),
                      n2cum = cumsum(n_2dose)),
             aes(t, n2cum, col = age_group)) +
  geom_step() + geom_vline(xintercept = c(12*7, 12*7+24), linetype = "dashed")

g3 <- ggplot(t1 %>% group_by(age_group) %>%
               mutate(n1cum = cumsum(n_1dose),
                      n2cum = cumsum(n_2dose)),
             aes(t, weighted_infection_efficacy, col = age_group)) +
  geom_step() + geom_vline(xintercept = c(12*7, 12*7+24), linetype = "dashed")

gg1 <- cowplot::plot_grid(g1,g2,g3)


# and now using default function
N <- 10000
t1 <- weighted_efficacy(iso3c = iso3c,
                        N = N,
                        maxt = 125,
                        doses_per_day = as.integer((first_doses + second_doses)*(N/sum(squire::get_population(iso3c = "BGD")$n))),
                        dose_period = 9 * 7,
                        v1v2 = 28,
                        prioritisation_matrix = nimue::strategy_matrix("Elderly", max_coverage = 0.8),
                        d2_prioritise = rep(TRUE, 17),
                        infection_efficacy = infection_efficacy,
                        disease_efficacy = disease_efficacy)

g1 <- ggplot(t1 %>% group_by(age_group) %>%
               mutate(n1cum = cumsum(n_1dose),
                      n2cum = cumsum(n_2dose)),
             aes(t, n1cum, col = age_group)) +
  geom_step() + geom_vline(xintercept = c(12*7, 12*7+24), linetype = "dashed")

g2 <- ggplot(t1 %>% group_by(age_group) %>%
               mutate(n1cum = cumsum(n_1dose),
                      n2cum = cumsum(n_2dose)),
             aes(t, n2cum, col = age_group)) +
  geom_step() + geom_vline(xintercept = c(12*7, 12*7+24), linetype = "dashed")

g3 <- ggplot(t1 %>% group_by(age_group) %>%
               mutate(n1cum = cumsum(n_1dose),
                      n2cum = cumsum(n_2dose)),
             aes(t, weighted_infection_efficacy, col = age_group)) +
  geom_step() + geom_vline(xintercept = c(12*7, 12*7+24), linetype = "dashed")

gg2 <- cowplot::plot_grid(g1,g2,g3)
````

Reply:
Yes - agreed also came to this realisation yesterday evening!